src/csharp/Intel/Iced/Intel/Extended/ExtendedEncoder.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Iced.Intel.BlockEncoderInternal;
+
+namespace Iced.Intel {
+	public sealed partial class ExtendedEncoder {
+
+		readonly CodeWriter _writer;
+		ulong _internalRip;
+		Label _label;
+		readonly List<Label> _labels;
+
+		private ExtendedEncoder(CodeWriter writer, int bitness) {
+			Debug.Assert(bitness == 16 || bitness == 32 || bitness == 64);
+			if (writer is null)
+				ThrowHelper.ThrowArgumentNullException_writer();
+			Bitness = bitness;
+			_writer = writer;
+			_labels = new List<Label>();
+			_label = CreateLabel();
+		}
+
+		public int Bitness { get; }
+
+		public static ExtendedEncoder Create(int bitness, CodeWriter writer) {
+			if (writer == null) throw new ArgumentNullException(nameof(writer));
+			switch (bitness) {
+			case 16:
+			case 32:
+			case 64:
+				return new ExtendedEncoder(writer, bitness);
+			default:
+				throw new ArgumentOutOfRangeException(nameof(bitness));
+			}
+		}
+
+		public Label CreateLabel(string name = null) {
+			var label = new Label(name, new InstructionBlock(_writer, new List<Instruction>(), _internalRip));
```

Comment:
I assume you'll copy `_internalRip` to a branch's NearBranch64 property. If so, you should initialize _internalRip to a non-zero value so it's unique (all the other instructions will have an "IP" of 0)

Reply:
Oh right, I over complicated with multiple InstructionBlock, as I haven't looked at how jumps are handled I thought they would be cleanly handle between instructionblocks, but it seems that the system will handle rip if I assign them (that's originally what I did)
Good, will fix this.

Reply:
There might be need of multiple blocks in different locations, especially for something like labels.

```
//Memory location: 0x40000:
var lbl1 = CreateLabel();
a.cmp(eax,100);
a.jbe(lbl1);
a.inc(eax);
a.jmp(lbl2)
a.label(lbl1);
a.xor(eax,eax)
a.jmp(lbl2)
```

```
//Memory location: 0x60000;
var lbl2 = CreateLabel();
a.jmp(lbl1)
a.label(lbl2);
```

These are two blocks that rely on each other label's RIP but both are unknown until encoded. So there are two solutions:

a) Replace the unknown label with the highest possible value and replace it after when you know the actual value. This is very problematic I guess. For all kind of jumps it changes the memory size of the instruction and missing bytes need to be replaced with nops.
b) Add support for the defining new rip locations in syntax directly for example:

```
var lbl1 = CreateLabel();
a.location(0x40000) //<-- or whatever you want to call this
a.cmp(eax,100);
a.jbe(lbl1);
a.inc(eax);
a.jmp(lbl2)
a.label(lbl1);
a.xor(eax,eax)
a.jmp(lbl2)
a.location(0x60000)
var lbl2 = CreateLabel();
a.jmp(lbl1)
a.label(lbl2);
```

Then it can be encoded all together. I would really appreciate if you can take care of something like this in the designing process.

Reply:
So branches are implemented with a single InstructionBlock and it seems to work well.
Next remaining work will be to support them from memory operand as well.